nvflare/app_common/hub/hub_controller.py

```diff
@@ -0,0 +1,490 @@
+            fl_ctx=fl_ctx,
+            dynamic_targets=True,
+            abort_signal=abort_signal,
+        )
+        if abort_signal.triggered:
+            return None
+
+        return task.get_prop(self._PROP_LAST_RESULT)
+
+    def _process_relay_result(self, client_task: ClientTask, fl_ctx: FLContext):
+        # submitted shareable is stored in client_task.result
+        # we need to update task.data with that shareable so the next target
+        # will get the updated shareable
+        task = client_task.task
+        current_round = task.get_prop(AppConstants.CURRENT_ROUND)
+        task.set_prop(self._PROP_LAST_RESULT, client_task.result)
+
+        task_data = client_task.result
+        shareable_generator = task.get_prop(self._PROP_SHAREABLE_GEN)
+        if shareable_generator:
+            # turn received result (a Shareable) to learnable (i.e. weight diff => weight)
+            learnable = shareable_generator.shareable_to_learnable(client_task.result, fl_ctx)
+
+            # turn the learnable to task data for the next leg (i.e. weight Learnable to weight Shareable)
+            task_data = shareable_generator.learnable_to_shareable(learnable, fl_ctx)
+
+        if current_round:
+            task_data.set_header(AppConstants.CURRENT_ROUND, current_round)
+        task.data = task_data
+        client_task.result = None
+
+
+class HubController(Controller):
+    def __init__(
+        self,
+        pipe_id: str,
+        task_wait_time=None,
+        task_data_poll_interval: float = 0.1,
+    ):
+        Controller.__init__(self)
+
+        check_positive_number("task_data_poll_interval", task_data_poll_interval)
+        check_str("pipe_id", pipe_id)
+        if task_wait_time is not None:
+            check_positive_number("task_wait_time", task_wait_time)
+
+        self.pipe_id = pipe_id
+        self.operator_descs = None
+        self.task_wait_time = task_wait_time
+        self.task_data_poll_interval = task_data_poll_interval
+        self.pipe = None
+        self.pipe_monitor = None
+        self.run_ended = False
+        self.task_abort_signal = None
+        self.current_task_name = None
+        self.current_task_id = None
+        self.current_operator = None
+        self.builtin_operators = {OperatorMethod.BROADCAST: BcastOperator(), OperatorMethod.RELAY: RelayOperator()}
+        self.project_name = ""
+
+    def start_controller(self, fl_ctx: FLContext) -> None:
+        self.project_name = fl_ctx.get_identity_name()
+
+        # get operators
+        engine = fl_ctx.get_engine()
+        job_id = fl_ctx.get_job_id()
+        workspace = engine.get_workspace()
+        app_config_file = workspace.get_server_app_config_file_path(job_id)
+        with open(app_config_file) as file:
+            app_config = json.load(file)
+            self.operator_descs = app_config.get(OperatorConfigKey.OPERATORS, {})
+            self.log_debug(fl_ctx, f"Got operator descriptions: {self.operator_descs}")
```

Comment:
Where do we expect people to define these "operator_descs"?

In "config_fed_server.json"? Or when they create the "Task" object.